boot/bootutil/src/swap_move.c

```diff
@@ -0,0 +1,480 @@
+    return BOOT_STATUS_SOURCE_NONE;
+}
+
+/*
+ * "Moves" the sector located at idx - 1 to idx.
+ */
+static void
+boot_move_sector_up(int idx, uint32_t sz, struct boot_loader_state *state,
+        struct boot_status *bs, const struct flash_area *fap_pri,
+        const struct flash_area *fap_sec)
+{
+    uint32_t new_off;
+    uint32_t old_off;
+    int rc;
+
+    /*
+     * FIXME: assuming sectors of size == sz, a single off variable
+     * would be enough
+     */
+
+    /* Calculate offset from start of image area. */
+    new_off = boot_img_sector_off(state, BOOT_SECONDARY_SLOT, idx);
+    old_off = boot_img_sector_off(state, BOOT_SECONDARY_SLOT, idx - 1);
+
+    if (bs->idx == BOOT_STATUS_IDX_0) {
+        rc = boot_erase_trailer_sectors(state, fap_pri);
+        assert(rc == 0);
+
+        rc = boot_status_init(state, fap_pri, bs);
+        assert(rc == 0);
+
+        rc = boot_erase_trailer_sectors(state, fap_sec);
+        assert(rc == 0);
+    }
+
+    rc = boot_erase_region(fap_sec, new_off, sz);
+    assert(rc == 0);
+
+    rc = boot_copy_region(state, fap_sec, fap_sec, old_off, new_off, sz);
+    assert(rc == 0);
+
+    rc = boot_write_status(state, bs);
+
+    bs->idx++;
+    BOOT_STATUS_ASSERT(rc == 0);
+}
+
+static void
+boot_swap_sectors(int idx, uint32_t sz, struct boot_loader_state *state,
+        struct boot_status *bs, const struct flash_area *fap_pri,
+        const struct flash_area *fap_sec)
+{
+    uint32_t pri_off;
+    uint32_t sec_off;
+    uint32_t sec_up_off;
+    int rc;
+
+    sec_up_off = boot_img_sector_off(state, BOOT_SECONDARY_SLOT, idx);
+    sec_off = boot_img_sector_off(state, BOOT_SECONDARY_SLOT, idx - 1);
+    pri_off = boot_img_sector_off(state, BOOT_PRIMARY_SLOT, idx - 1);
+
+
+    if (bs->state == BOOT_STATUS_STATE_0) {
+        rc = boot_erase_region(fap_sec, sec_off, sz);
+        assert(rc == 0);
+
+        rc = boot_copy_region(state, fap_pri, fap_sec, pri_off, sec_off, sz);
+        assert(rc == 0);
+
+        rc = boot_write_status(state, bs);
+        bs->state = BOOT_STATUS_STATE_1;
+        BOOT_STATUS_ASSERT(rc == 0);
+    }
+
+    if (bs->state == BOOT_STATUS_STATE_1) {
+        rc = boot_erase_region(fap_pri, pri_off, sz);
+        assert(rc == 0);
+
+        rc = boot_copy_region(state, fap_sec, fap_pri, sec_up_off, pri_off, sz);
+        assert(rc == 0);
+
+        rc = boot_write_status(state, bs);
+        bs->idx++;
+        bs->state = BOOT_STATUS_STATE_0;
+        BOOT_STATUS_ASSERT(rc == 0);
+    }
+}
+
+/*
+ * While starting a new swap the primary slot is initialized from the data
+ * in the secondary slot. If this is detected as a revert, no information in
+ * the secondary slot exists; so, if the swap is interrupted at the "right"
+ * time the swap will not revert. This function handles the issue by faking
+ * a permanent upgrade in case a reset occurs.
+ */
+void
+fixup_revert(const struct boot_loader_state *state, struct boot_status *bs,
```

Comment:
Can you add more explanation for this? E.g. what is the "right" time?

Reply:
It might be a bad wording; maybe it should be called "wrong" instead of "right". The idea is that there is a short period of time during the beginning of a revert where metadata might be lost because status in slot0 must be erased and status in slot1 was already previously erased. If a reset occurs at this time the swap breaks. The `fixup...` routine just saves temporarily the data in slot1 (and it will be erased again moments later), but it is saved long enough to walk a few extra steps without loosing information.

Reply:
Thanks, I got what you meant by "right" ;). Your comment gives good context which I think should be added to the function doc comment.

Reply:
I did not copy the previous answer, but hopefully I improved the text with: 29ec1ab